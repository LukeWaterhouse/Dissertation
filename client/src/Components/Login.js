import { useState, useContext } from 'react'
import axios from 'axios'
import UserContext from './UserContext'
import '../Css/Register.css'
import { useHistory } from 'react-router-dom'

function Login() {
  const [userName, setUserName] = useState('')
  const [password, setPassword] = useState('')
  const [passwordError, setPasswordError] = useState(false)
  const [userNameError, setUserNameError] = useState(false)

  const [isFieldBlank, setIsFieldBlank] = useState(false)

  const user = useContext(UserContext)
  let history = useHistory()

  function loginUser(e) {
    e.preventDefault()

    if (userName == '' || password == '') {
      setIsFieldBlank(true)
      setPasswordError(false)
      setUserNameError(false)
    } else {
      setIsFieldBlank(false)

      const data = { userName, password }
      axios
        .post('http://localhost:5000/login', data, { withCredentials: true })
        .then((response) => {
          console.log(response.data)
          console.log('Posted!')

          if (response.data === 'noUserName') {
            setPasswordError(false)
            setUserNameError(true)
          } else {
            user.setUserName(response.data.userName)
            setUserName('')
            setPassword('')
            setPasswordError(false)
            setUserNameError(false)
            history.push('/homepage')
          }
        })
        .catch(() => {
          setPasswordError(true)
          setUserNameError(false)
        })
    }
  }

  return (
    <div className="container">
      <h3>Welcome to Data House</h3>
      <h4>Login</h4>
      <form action="" onSubmit={(e) => loginUser(e)}>
        <div>
          {isFieldBlank && (
            <div style={{ color: 'red' }}>Please fill out both fields!</div>
          )}

          {passwordError && (
            <div style={{ color: 'red' }}>Password incorrect!</div>
          )}
          {userNameError && (
            <div style={{ color: 'red' }}>Username does not exist!</div>
          )}
        </div>

        <div style={{ color: 'white' }}>Username: </div>
        <input
          type="text"
          id="usernameInput"
          placeholder="Username"
          value={userName}
          onChange={(e) => setUserName(e.target.value)}
        />
        <br />
        <div style={{ color: 'white', paddingTop: '10px' }}>Password</div>
        <input
          type="password"
          id="passwordInput"
          placeholder="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
        <br />
        <div
          style={{
            paddingTop: '20px',
            display: 'flex',
            justifyContent: 'center'
          }}
        >
          <button className="btn-1" type="submit">
            LOGIN
          </button>
        </div>
      </form>
    </div>
  )
}

export default Login
