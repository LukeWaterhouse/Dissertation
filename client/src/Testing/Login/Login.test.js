import React from 'react'
import ReactDOM from 'react-dom'
import { render, cleanup, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Login from '../../Components/Login'
import renderer from 'react-test-renderer'

afterEach(cleanup)

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Login></Login>, div)
  screen.debug
})

it('error message shows if you click login without entering info', () => {
  render(<Login />)
  userEvent.click(screen.getByRole('button'))
  expect(screen.getByText(/Please fill out both fields!/i)).toBeInTheDocument()
})

it('error message shows if you click login without entering info', () => {
  render(<Login />)
  userEvent.type(screen.getByRole('textbox'), 'sampleUsername')
  userEvent.type(screen.getByPlaceholderText(/password/i), 'samplePassword')
  userEvent.click(screen.getByRole('button'))

  expect(screen.queryByText('Please fill out both fields!')).toBeNull()
})

it('login matches snapshot', () => {
  const tree = renderer.create(<Login />).toJSON()
  expect(tree).toMatchSnapshot()
})
