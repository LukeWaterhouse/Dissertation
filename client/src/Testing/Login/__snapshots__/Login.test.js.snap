// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login matches snapshot 1`] = `
<div
  className="container"
>
  <h3>
    Welcome to GeckoSpot
  </h3>
  <h4>
    Login
  </h4>
  <form
    action=""
    onSubmit={[Function]}
  >
    <div />
    <div
      style={
        Object {
          "color": "white",
        }
      }
    >
      Username: 
    </div>
    <input
      id="usernameInput"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <br />
    <div
      style={
        Object {
          "color": "white",
          "paddingTop": "10px",
        }
      }
    >
      Password
    </div>
    <input
      id="passwordInput"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <br />
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "paddingTop": "20px",
        }
      }
    >
      <button
        className="btn-1"
        type="submit"
      >
        LOGIN
      </button>
    </div>
  </form>
</div>
`;
